---
title: "Finding Differentially Expressed Genes in the Kennerdell et al. (2018) RNA-seq data."
author: "Jason Kennerdell"
date: "11/15/2017"
output: pdf_document
---
##Synopsis:
  Here I am calling genes as different differentially expressed in the *E(z)* mutant *Drosophila* brains, and genes that are differentially expressed with age. The csv files are used to create the Supplemental Data 1 spreadsheet.

```{r, "needed packages", include=FALSE, warning=FALSE, message=FALSE}
library(DESeq2); library(ggplot2); library(grid); library(gridExtra); library(genefilter)
library(biomaRt); library(gplots)
```

### Input the data and metadata:
```{r, "input sampleTable", warning=FALSE, message=FALSE}
inpath <- "~/Desktop/brain"
outpath <- "~/Desktop/brain/CalledDE"
setwd(inpath)
library(RColorBrewer)
colSet <- brewer.pal(12, "Paired")
files <- list.files()
htseq_files <- files[grepl("^JKL.*txt$", files)]
EzNames <- read.csv("EzSampleNames.csv")
PclNames <- read.csv("PclSampleNames.csv")
sampleNames <- rbind(EzNames, PclNames)
sampleTable <- data.frame(fileName = htseq_files,
                          stringsAsFactors=FALSE)
sampleTable$Library <- gsub("-counts.txt", "", sampleTable$fileName)
sampleTable$Library <- gsub("b", "", sampleTable$Library)
sampleTable$seq.batch <- ifelse(grepl("b", sampleTable$fileName), "B", "A")
sampleTable$seq.batch <- paste(sampleTable$Library, sampleTable$seq.batch)
sampleTable <- merge(sampleTable, sampleNames[,1:3], by = "seq.batch")
sampleTable$genotype <- gsub("-.*$", "", sampleTable$Sample)
sampleTable$genotype <- gsub("JKLY.... ", "", sampleTable$genotype)
sampleTable$genotype <- gsub("\\[1118\\]", "", sampleTable$genotype)
sampleTable$genotype <- factor(sampleTable$genotype, levels = c("w", "Ez"))
sampleTable <- sampleTable[!is.na(sampleTable$genotype),]
sampleTable$Temp <- gsub("[A-Z, a-z, -]", "", sampleTable$Sample)
sampleTable$Temp <- gsub("^11.*$", "25", sampleTable$Temp)
sampleTable$Temp <- factor(sampleTable$Temp, levels = c("25", "29"))
sampleTable$age <- gsub("JKLY.*$", "3d", sampleTable$Sample)
sampleTable$age <- gsub("^[a-zA-z].*$", "20d", sampleTable$age)
sampleTable$age <- factor(sampleTable$age, levels = c("3d", "20d"))
sampleTable$condition <- paste(sampleTable$genotype, sampleTable$Temp, sampleTable$age, sep = "-")
sampleTable$color <- c(rep(colSet[7], 4), rep(colSet[9], 2), 
                       rep(colSet[8], 2), rep(colSet[10], 2), rep(colSet[8], 2), rep(colSet[10], 2), 
                       rep(colSet[8], 2), rep(colSet[10], 2), rep(colSet[8], 2), 
                       rep(colSet[1], 2),
                       rep(colSet[10], 2), rep(colSet[8], 2), rep(colSet[10], 2),
                       rep(colSet[1], 4), rep(colSet[7], 2), rep(colSet[9], 4))
sampleTable
```

### Set up the statistical model to test for Differentially Expressed genes in E(z) mutants:
```{r, "design"}
design <- formula(~ Temp + age + genotype)
```

### DESeq2 Statistics
```{r, "statistics", message=FALSE}
dds <- DESeqDataSetFromHTSeqCount(sampleTable = sampleTable, directory = inpath, design = design)
# Combine the technical replicates (different runs) by adding the count
# totals for each gene across the two runs:
dds <- collapseReplicates(dds, groupby=dds$Library, run = dds$batch)
dds <- DESeq(dds)
# What does the data look like?
head(assay(dds))
# What are the columns?
colData(dds)
resultsEz <- results(dds, alpha=0.05) # this gives identical for E(z) vs wt contrast
resultsEz$ensembl <- rownames(resultsEz)
summary(resultsEz)
resultsAge <- results(dds, alpha=0.05, contrast=c("age", "20d", "3d"))
resultsAge$ensembl <- rownames(resultsAge)
summary(resultsAge)
```

### Add Annotations
```{r}
# Add usefull gene names:
library(biomaRt)
mart = useMart("ENSEMBL_MART_ENSEMBL", host="aug2017.archive.ensembl.org")
#listDatasets(mart)
mart = useMart("ENSEMBL_MART_ENSEMBL", host="aug2017.archive.ensembl.org",
               dataset = "dmelanogaster_gene_ensembl")

# For the Ez gene list:
genemap <- getBM(attributes = c("ensembl_gene_id", "entrezgene", "external_gene_name", 
                                "flybasecgid_gene"), filters = "ensembl_gene_id", 
                 values = resultsEz$ensembl, mart = mart)
idx <- match(resultsEz$ensembl, genemap$ensembl_gene_id)
resultsEz$entrez <- genemap$entrezgene[idx]
resultsEz$geneSymbol <- genemap$external_gene_name[idx]
resultsEz$cg <- genemap$flybasecgid_gene[idx]
write.csv(as.data.frame(resultsEz), file = paste(outpath, "EzContrast_Benjp05.csv", sep="/"))
# For the Aging gene list:
genemap <- getBM(attributes = c("ensembl_gene_id", "entrezgene", "external_gene_name", 
                                "flybasecgid_gene"), filters = "ensembl_gene_id", 
                 values = resultsAge$ensembl, mart = mart)
idx <- match(resultsAge$ensembl, genemap$ensembl_gene_id)
resultsAge$entrez <- genemap$entrezgene[idx]
resultsAge$geneSymbol <- genemap$external_gene_name[idx]
resultsAge$cg <- genemap$flybasecgid_gene[idx]
write.csv(as.data.frame(resultsAge), file = paste(outpath, "AgeContrast_Benjp05.csv", sep="/"))
```

## Additional Diagnostic Plots
```{r, include=F, eval=F}
# Cooks Distances for each sample:
boxplot(log10(assays(dds)[["cooks"]]), range=0, las=2)
# Dispersion Plot:
plotDispEsts(dds)
```

### Print out gene names of called diferentially expressed genes:
```{r, include = F}
subset(resultsEz, padj < 0.05 & log2FoldChange > 0.5)$geneSymbol
subset(resultsEz, padj < 0.05 & log2FoldChange < -0.5)$geneSymbol
subset(resultsAge, padj < 0.05 & log2FoldChange > 0.5)$geneSymbol
subset(resultsAge, padj < 0.05 & log2FoldChange < -0.5)$geneSymbol
```

## How about Bonferroni corrected data?
```{r}
# For Ez Contrast:
resultsEzBonf <- results(dds, alpha = 0.05, pAdjustMethod = "bonferroni")
resultsEzBonf$ensembl <- rownames(resultsEzBonf)
idxEzBonf <- match(resultsEzBonf$ensembl, genemap$ensembl_gene_id)
resultsEzBonf$geneSymbol <- genemap$external_gene_name[idxEzBonf]
write.csv(as.data.frame(resultsEzBonf), file = paste(outpath, "EzContrast_Bonfp05.csv", sep="/"))
# For Aging Contrast:
resultsAgeBonf <- results(dds, alpha=0.05, contrast=c("age", "20d", "3d"), pAdjustMethod = "bonferroni")
resultsAgeBonf$ensembl <- rownames(resultsAgeBonf)
idxAgeBonf <- match(resultsAgeBonf$ensembl, genemap$ensembl_gene_id)
resultsAgeBonf$geneSymbol <- genemap$external_gene_name[idxAgeBonf]
write.csv(as.data.frame(resultsAgeBonf), file = paste(outpath, "AgeContrast_Bonfp05.csv", sep="/"))
```

### Prepare lists of DE genes:
```{r}
datEz <- as.data.frame(resultsEz)
datEz$Bonf.correction <- as.data.frame(resultsEzBonf)$padj
calledEz <- subset(datEz, padj < 0.05 & abs(log2FoldChange) > 0.5)
write.csv(calledEz, file = paste(outpath, "EzCalledDE.csv", sep="/"))
datAge <- as.data.frame(resultsAge)
datAge$Bonf.correction <- as.data.frame(resultsAgeBonf)$padj
calledAge <- subset(datAge, padj < 0.05 & abs(log2FoldChange) > 0.5)
write.csv(calledAge, file = paste(outpath, "AgeCalledDE.csv", sep="/"))
```


